### PR TITLE
Bug 2008539: Try another registry if blob is not found

### PR DESCRIPTION
**Pull request by dmage**
ServeBlob uses only the first connected registry, but it does not check
that the registry has needed data. If the blob is absent of this
registry or the registry rejects the request, the client can safely make
another attempt with another registry.

### CODE REVIEW

**Comment by xiuwang** [comment]
Since cluster-bot can't launch cluster from library-go repo, I just reviewed the code.
I will test this bug after pr merged.

File: pkg/image/registryclient/client_mirrored.go
Comment (smarterclayton) [inline pkg/image/registryclient/client_mirrored.go:479]: Yikes.  What are the performance characteristic changes to this?

This changes every trivial serve blob into a stat?

Comment (dmage) [reply pkg/image/registryclient/client_mirrored.go:479]: In most cases it won't causes additional HTTP requests.

https://github.com/distribution/distribution/blob/1563384b69df9376389fe45ce949173a6383770a/registry/client/transport/http_reader.go#L66-L93

It forces the client to establish an HTTP connection and this connection will be reused by next reads. It won't be reused only if the caller decides to Seek before Read.

(Btw, we increased the number of requests by https://github.com/openshift/library-go/pull/1084, see https://github.com/openshift/image-registry/pull/278/commits/d7eeb86d675641dfcda4ba2dab0ef2f3d8e8e7b7; perhaps we need to check why)

Comment (dmage) [reply pkg/image/registryclient/client_mirrored.go:479]: (`r.Blobs(ctx)` creates a BlobService with an internal cache, as we recreate BlobService on every call, it does not cache anything)

Comment (dmage) [reply pkg/image/registryclient/client_mirrored.go:479]: (https://github.com/openshift/library-go/compare/master...dmage:cache-blobstore addresses the problem with the extra request, but as it is not related to the BZ, perhaps it should be a follow-up PR?)

File: pkg/image/registryclient/client_mirrored.go
Comment (soltysh) [inline pkg/image/registryclient/client_mirrored.go:219]: Why in those particular cases we would return instead of falling back to other registries?

**Comment by dmage** [comment]
@soltysh @smarterclayton the PR is ready for the second review, ptal

**Comment by dmage** [comment]
@soltysh @smarterclayton please take a look and tag it

**Comment by dmage** [comment]
We need to update this library in `image-registry` so that the image-registry can use all mirrors.
We can update this library in `oc` so that it can use all mirrors and has better error messages when it mirrors images.

**Review by soltysh — APPROVED** [review]
/lgtm
@smarterclayton has approval rights

**Comment by deads2k** [comment]
> We need to update this library in `image-registry` so that the image-registry can use all mirrors.
> We can update this library in `oc` so that it can use all mirrors and has better error messages when it mirrors images.

this sounds reasonable to me.  The additional call in error cases looks like it degrades the already degraded cases.

/approve

**Pull request merged**
